### PR TITLE
add authentication and authorization to translation api

### PR DESCRIPTION
**Pull request by blaisefonseca**

### CODE REVIEW

File: app/api/database-table.js [hunk 1]
```diff
@@ -1,41 +1,79 @@
...
+            
+            let q = req.decryptedInfo;
+            
+            let databaseTable = _.find(databaseTables, {name:req.params.table})
+            let files = await getUpdatesFiles(databaseTable, q);
+
+            if(files && files.length > 0){
                 let compressedFiles = await zipFile(req.params.table, files);
                 let zip4 = new EasyZip();
                 return zip4.batchAdd(compressedFiles, function(){
```
Comment (randyhoulahan) [inline app/api/database-table.js]: should this be streamed?

File: app/api/database-table.js [hunk 2]
```diff
@@ -1,41 +1,79 @@
...
-            res.status(200).send(files);
+            res.status(200).send('No files found for download');
         }
         catch(err) {
+            winston.error(err)
             res.status(500).send('Unknown error occurred');
         };
     }
 
+    async function authorized(req, res, next){
```
Comment (randyhoulahan) [inline app/api/database-table.js]: is async needed here?

Comment (sbilodeau) [reply app/api/database-table.js]: no

File: app/api/database-table.js [hunk 3]
```diff
@@ -1,41 +1,79 @@
...
+        try{
+
+            if(!req.user || !authenticate.isInRole(req.user, ['Administrator', 'oasisArticleEditor'])){
+                return res.status(401).send('You are not authorized to access this resource');
```
Comment (sbilodeau) [inline app/api/database-table.js]: 403 (~401~ is for not authenticated)

Comment (sbilodeau) [inline app/api/database-table.js]:
```suggestion
                return res.status(403).send('You are not authorized to access this resource');
```

File: app/api/database-table.js [hunk 4]
```diff
@@ -1,41 +1,79 @@
...
+            }
+            
+            next();
+        }
+        catch(err) {
+            res.status(500).send('Unknown error occurred');
```
Comment (sbilodeau) [inline app/api/database-table.js]: Should we log the error to the console?

File: app/api/git-query.js [hunk 5]
```diff
@@ -1,36 +1,81 @@
...
+            let q = req.query;
+            
+            let repositoryName = req.params.repository;
+            let files = await getUpdatesFiles(repositoryName, q.branch, q.date, q.ignoreFiles, q.allowedExtenstions);
+
+            res.status(200).send(files);
         }
-        res.status(200).send(files);
+        catch(err) {
+            res.status(500).send('Unknown error occurred');
```
Comment (sbilodeau) [inline app/api/git-query.js]: Same: should we log the error to the console

File: app/api/git-query.js [hunk 6]
```diff
@@ -1,36 +1,81 @@
...
+            winston.error(err)
+            res.status(500).send('Unknown error occurred');
+        };
+    }
+
+    async function authorized(req, res, next){
+        try{
+
+            if(!req.user || !authenticate.isInRole(req.user, ['Administrator', 'oasisArticleEditor'])){
+                return res.status(401).send('You are not authorized to access this resource');
```
Comment (sbilodeau) [inline app/api/git-query.js]: 403 ~not 401~

https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses

File: app/api/signed-url.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+//
+//
+//============================================================
+function signedUrl (req, res) {
+
+      this.get = function(req, res) {
+        try {
+           
+            if(!req.user || req.user.anonymous)
+                return res.status(401).send('You are not authorized to access this resource');
```
Comment (sbilodeau) [inline app/api/signed-url.js]: This is the good usage of 401!

File: app/api/signed-url.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    this.isValid = function(req, res, next) {
+        try {
+            if(!req.query || !req.query.hash)
+                return res.status(401).send('Url not signed');
+                       
+            var details = encryption.decrypt(req.query.hash)
+            details = JSON.parse (details);
+            if(!details.expiry || details.expiry < new Date())
+                return res.status(401).send('Invalid signed url hash');
```
Comment (sbilodeau) [inline app/api/signed-url.js]: 403

File: app/views/translation/table.js [hunk 9]
```diff
@@ -17,28 +21,57 @@
...
+            });
+        }
 
         $scope.dowloadFiles = function(){
+
+
             var translation = $scope.translation;
 
-            var url = '/translation-api/database-table/'+translation.name +'?zip=true';
+            var url = baseUrl+'translation-api/database-table/'+translation.name ;
```
Comment (sbilodeau) [inline app/views/translation/table.js]:
```suggestion
            var url = baseUrl+'translation-api/database-table/'+ encodeURIComponent (translation.name);
```

File: app/views/translation/table.js [hunk 10]
```diff
@@ -17,28 +21,57 @@
...
             var filesForTranslation = _.map(_.filter($scope.translation.rows, {translate:true}), '_id');
 
             if(filesForTranslation.length > 0){
-                url += '&ids='+filesForTranslation.join('&ids=')
-                window.open(url,'_new');
+                $scope.gettingSignedUrl = true;
+                $http.get(url+'/signed-url', {params:{ids:filesForTranslation}})
+                .then(function(result){
+                    window.open(url+'/download?hash='+ result.data,'_new');
```
Comment (sbilodeau) [inline app/views/translation/table.js]:
```suggestion
                    window.open(url+'/download?hash='+ encodeURIComponent(result.data),'_new');
```

File: app/api/authentication.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (sbilodeau) [inline app/api/authentication.js]: We should think about making this file an external module one day for reuse in other projects (when day we need it comes)

File: app/api/config.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+﻿
+let config = {}
+
+config.api      = {};
+config.api.url  = process.env.API_URL || 'https://api.cbddev.xyz:443';
```
Comment (sbilodeau) [inline app/api/config.js]: May be we should trim the trailing / 
```suggestion
config.api.url  = (process.env.API_URL || 'https://api.cbddev.xyz:443').replace(/\/*$/, '')
```

File: app/api/config.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+
+config.encryptionPassword   = process.env.ENCRYPTION_PASSWORD
```
Comment (sbilodeau) [inline app/api/config.js]: can you throw an exception is missing or sort

File: app/views/translation/project.js [hunk 14]
```diff
@@ -1,7 +1,7 @@
 ﻿define(['app', 'lodash'], function (app, _) {
     return ['$scope', '$http', '$q', '$routeParams',
     function ($scope, $http, $q, $routeParams) {
-            var baseUrl = window.baseUrl||'/';
+            var baseUrl = $scope.baseUrl = window.baseUrl||'/';
```
Comment (sbilodeau) [inline app/views/translation/project.js]: trailing / is mandatory in this case.
may be you should ensure it's there

```suggestion
            var baseUrl = $scope.baseUrl = window.baseUrl.rerplace(/\/*$/, '') + '/';
```

File: app/views/translation/project.js [hunk 15]
```diff
@@ -46,5 +46,30 @@
 
                 }
             }
+
+            $scope.dowloadFiles = function(){
+               
+                if($scope.translation.previousRelease){
+
+                    var url = baseUrl+'translation-api/git/'+ $scope.project.name;
```
Comment (sbilodeau) [inline app/views/translation/project.js]:
```suggestion
                    var url = baseUrl+'translation-api/git/'+ encodeURIComponent($scope.project.name);
```

File: app/views/translation/project.js [hunk 16]
```diff
@@ -46,5 +46,30 @@
...
+                        ignoreFiles         : $scope.translation.ignoreFiles,
+                        allowedExtenstions  : $scope.translation.allowedExtenstions
+                    }
+                    if($scope.translation.allFiles)
+                        params.date = undefined;
+
+                    $scope.gettingSignedUrl = true;
+                    $http.get(url+'/signed-url', {params:params})
+                    .then(function(result){
+                        window.open(url+'/download?hash='+ result.data,'_new');
```
Comment (sbilodeau) [inline app/views/translation/project.js]:
```suggestion
                        window.open(url+'/download?hash='+ encodeURIComponent(result.data),'_new');
```

File: app/views/translation/table.js [hunk 17]
```diff
@@ -1,13 +1,17 @@
-﻿define(['app', 'lodash', 'json!views/translation/database-tables.json'], function (app, _, dbTables) {
+﻿define(['app', 'lodash', 'json!views/translation/database-tables.json','views/translation/directives/pagination'],
+ function (app, _, dbTables) {
     return ['$scope', '$http', '$q', '$routeParams',
     function ($scope, $http, $q, $routeParams) {
         var languages = [ 'ar', 'fr', 'es', 'ru', 'zh' ]
         $scope.baseUrl  = window.baseUrl||'/';
```
Comment (sbilodeau) [inline app/views/translation/table.js]: same
```suggestion
        $scope.baseUrl  = window.baseUrl.replace(/\/*$/,'')+'/';
```

File: app/api/config.js (hunk 13 above)
Comment (blaisefonseca) [reply app/api/config.js]: i am throwing exception in the encryption file

File: app/views/translation/project.js (hunk 14 above)
Comment (blaisefonseca) [reply app/views/translation/project.js]: while this is somehting that comes from out ENV variables, i will move it to .ejs to i dont have to do it in all files,, ` window.baseUrl = '<%=baseUrl.replace(/\/*$/, "") + "/" %>';`

**Pull request merged**